src/client/datascience/jupyter/kernels/kernelExecution.ts

```diff
@@ -268,11 +266,14 @@ export class KernelExecution implements IDisposable {
 
     @captureTelemetry(Telemetry.RestartKernel)
     @captureTelemetry(Telemetry.RestartJupyterTime)
-    private async restartExecution(_document: NotebookDocument, session: IJupyterSession): Promise<void> {
+    private async restartExecution(notebook: INotebook): Promise<void> {
         // Just use the internal session. Pending cells should have been canceled by the caller
-        return session.restart(this.interruptTimeout).catch((exc) => {
+        await notebook.session.restart(this.interruptTimeout).catch((exc) => {
             traceWarning(`Error during restart: ${exc}`);
         });
+
+        // Reinitialize the kernel after a session restart
+        await notebook.runInitialSetup();
```

Comment:
This was never required for Native notebooks, now we have code thats specific to interactive widow.
I think we should put an `if` condition for this and ensure this only runs for Intearctive window, as we have initialization for Native Notebooks already in the IKernel clas.

Reply:
This is required for native notebooks, and this is how restart used to work until 2 weeks ago. See https://github.com/microsoft/vscode-jupyter/pull/6826/files#diff-7d0d36cd88827937fc9458d80a807484935c9d2b5d63501a6ec98eb1e5351c16L178

Reply:
I don't see any old code that called `notebook.runInitialSetup`
That's new, Or an I missing something

Reply:
That's a new method I added to expose the initialization logic which used to run as part of INotebook.restartKernel

Reply:
> PR is to move the initialization logic to IKernel.initializeAfterStart. What do you think?

Yes, thats much better, that way all the initialization is in one place.